File: arrays/secondMaxVal.js
 /* Implement a function which returns the second largest element in an array of integers. */

 /* PEDAC */
 /*   in: array of integers */
 /*   out: integer, the second largest in input */
 /*       clarify: 'second largest' requires at least two distinct values; */
 /*                 duplicate values count only 1x */
 /*   examples: */
 /*     see tests below */

 /*   data structure: */
 /*     hash / (or JS Map) holding current values for largest and second largest */

 /*   algorithm: */
 /*     guard clause: return nil if input size < 2 */
 /*     init hash, largest: '-Infinity', second_largest: '-Inifinity' */

 /*     iterate over array: */
 /*       if current value <= second_largest */
 /*         next */
 /*       if current value > largest */
 /*         second_largest = largest */
 /*         largest = current value */
 /*       if current value > second_largest && current value < greatest */ 
 /*         second_largest = current value */

 /*         return hsh[:second_largest] if it is greater than -infinity, else return nil */

// Space: O(1) for the map object
// Time: O(N)

const assert = require('assert').strict;

const findSecondMax = (array) => {
  const values = new Map([ ['greatest', array[0]] ]);

  for (let value of array) {
    let max = values.get('greatest');
    let second = values.get('second') || -Infinity;

    if (value <= second) continue;

    if (value > max) {
      values.set('second', max);
      values.set('greatest', value);
    } else if ( (value > second) && (value < max) ) {
      values.set('second', value);
    }
  }

  return values.get('second');
};


// test_finds_target_at_last_idx
  assert.equal(findSecondMax([9,2,3,6]), 6)

// test_finds_target_at_first_idx
  assert.equal(findSecondMax([4,3,1,5,0]), 4)

// test_finds_target_in_middle
  assert.equal(findSecondMax([4,3,1,9,5,10]), 9)

// test_does_not_confuse_duplicates_with_different_value_rank
  assert.equal(findSecondMax([4,3,1,9,5,10,10]), 9)

// test_returns undefined_for_target_not_found
  assert.equal(findSecondMax([8,8,8,8]), undefined)

// test_returns undefined_for_target_not_found_empty_input
  assert.equal(findSecondMax([]), undefined)

// test_returns undefined_for_target_not_found_single_value_input
  assert.equal(findSecondMax([22]), undefined)

File: arrays/second_max_val.py
# Implement a function which returns the second largest element in an array of integers.

# in: array of int
# out: second largest value int

# examples: see tests

# algo: 
# create max dict with two keys, first and second; set each to -inf
# iterate over list:
    # if current val is > first set it to first and set second to value of previous first
    # else if current val is < first > second, set second to current val

# return second or None in case second is neg inf

# Time: O(N) | Space O(1)
def second_max(ints):
    max = {
            'first': -float('inf'),
            'second': -float('inf')
            }

    for i in ints:
        if i > max['first']:
            max['second'] = max['first']
            max['first'] = i
        elif max['first'] > i > max['second']:
            max['second'] = i

    if max['second'] > -float('inf'):
        return max['second'] 
    else:
        return None


class TestSecondMax:
    def test_target_at_end(self):
        assert second_max([9,2,3,6]) == 6

    def test_target_at_start(self):
        assert second_max([4,3,1,5,0]) == 4

    def test_target_in_mid(self):
        assert second_max([1,2,3,6,4,1,9]) == 6

    def test_duplicate_max(self):
        assert second_max([1,2,3,3,2]) == 2

    def test_all_one_value1(self):
        assert second_max([1,1,1,1,1]) == None

    def test_all_one_value2(self):
        assert second_max([1]) == None
